More scripting

lukedirtwalker opened the pull request:
Some more commits, so that you will not get bored next week :)
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/dimitar-asenov/Envision/pull/143%23discussion_r38788697%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/143%23discussion_r38789367%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/143%23discussion_r38790190%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/143%23discussion_r38790438%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/143%23discussion_r38790550%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/143%23discussion_r38790644%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/143%23discussion_r38790746%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/143%23discussion_r38791172%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/143%23discussion_r38792935%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/143%23discussion_r38793164%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/143%23discussion_r39378490%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/143%23discussion_r39380331%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/143%23discussion_r39383850%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/143%23discussion_r39383876%22%2C%20%22https%3A//github.com/dimitar-asenov/Envision/pull/143%23discussion_r39384124%22%5D%2C%20%22comments%22%3A%20%7B%22Pull%20778b5e60e4dc3272a27fe47294199c531969399b%20InformationScripting/src/queries/QueryRegistry.h%2039%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/143%23discussion_r38790644%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22I%20have%20an%20idea%2C%20but%20I%27d%20like%20to%20hear%20from%20you%2C%20how%20exactly%20you%20imagine%20we%27ll%20use%20this%20registry.%22%2C%20%22created_at%22%3A%20%222015-09-04T20%3A25%3A28Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22Well%20it%20is%20used%20to%20get%20all%20known%20queries%20in%20the%20system.%20Or%20what%20do%20you%20mean%20how%20we%20will%20use%20it%3F%22%2C%20%22created_at%22%3A%20%222015-09-04T20%3A52%3A46Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3530769%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/lukedirtwalker%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/queries/QueryRegistry.h%3AL1-70%22%7D%2C%20%22Pull%20778b5e60e4dc3272a27fe47294199c531969399b%20InformationScripting/src/queries/SubstractNodesOperator.cpp%2013%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/143%23discussion_r38791172%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Hmm%2C%20I%27m%20not%20entirely%20convinced%20what%27s%20the%20correct%20operation%20of%20the%20subtract%20operator.%20Strictly%20speaking%2C%20you%20shouldn%27t%20extend%20the%20B%20set%20before%20subtracting%20it%20from%20A.%20But%20on%20the%20other%20hand%2C%20it%20might%20be%20useful%2C%20in%20which%20case%20doing%20it%20here%20seems%20OK%2C%20why%20is%20it%20not%20OK%3F%20Maybe%20it%20should%20be%20an%20option.%20I%20can%20also%20imagine%20that%20there%20might%20be%20a%20more%20complex%20conditions%20that%20determines%20whether%20we%20should%20subtract%20something.%22%2C%20%22created_at%22%3A%20%222015-09-04T20%3A31%3A45Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/queries/SubstractNodesOperator.cpp%3AL35-43%22%7D%2C%20%22Pull%20778b5e60e4dc3272a27fe47294199c531969399b%20InformationScripting/src/dataformat/TupleSet.h%2017%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/143%23discussion_r38789367%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22I%27m%20not%20quite%20happy%20with%20the%20name%20of%20this%20method.%20Let%27s%20call%20it%20something%20more%20descriptive%2C%20such%20as%20%60addPropertiesAsTuples%60%20or%20some%20other%20better%20name.%22%2C%20%22created_at%22%3A%20%222015-09-04T20%3A08%3A05Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/dataformat/TupleSet.h%3AL55-67%22%7D%2C%20%22Pull%20888787134efffdbbabcc4c21dd0cc7479cb578d3%20InformationScripting/src/parsing/QueryBuilder.cpp%20139%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/143%23discussion_r39383850%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22The%20past%20tense%20of%20split%20is%20split.%22%2C%20%22created_at%22%3A%20%222015-09-14T11%3A33%3A13Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/parsing/QueryBuilder.cpp%3AL1-210%22%7D%2C%20%22Pull%20778b5e60e4dc3272a27fe47294199c531969399b%20InformationScripting/src/parsing/QueryBuilder.cpp%20207%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/143%23discussion_r38790190%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22OK%2C%20I%27m%20not%20even%20going%20to%20pretend%20that%20I%20really%20tried%20to%20understand%20the%20parser.%20I%20looked%20at%20the%20code%20briefly%20and%20kept%20thinking%3A%20why%20is%20this%20so%20complicated%3F%20I%20must%20be%20missing%20something%20major%2C%20because%20I%20thought%20that%20our%20command%20expressions%20are%20so%20simple%2C%20we%20can%20literally%20use%20a%20few%20regular%20expressions%20to%20implement%20the%20parser.%20What%20are%20you%20thoughts%20on%20this%2C%20do%20we%20really%20need%20all%20the%20complex%20stuff%20above%3F%20If%20so%2C%20could%20you%20try%20to%20elaborate%20why%20that%20is%3F%22%2C%20%22created_at%22%3A%20%222015-09-04T20%3A19%3A23Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22Hm%20have%20you%20looked%20at%20the%20grammar%20it%20should%20be%20able%20to%20parse%20in%20the%20header%20file%3F%20And%20after%20looking%20at%20it%20are%20you%20still%20sure%20we%20can%20solve%20it%20with%20regular%20expressions%3F%20If%20so%20I%20will%20be%20happy%20to%20redo%20it%20with%20regexes.%22%2C%20%22created_at%22%3A%20%222015-09-04T20%3A55%3A14Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3530769%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/lukedirtwalker%22%7D%7D%2C%20%7B%22body%22%3A%20%22Ok%2C%20I%20think%20we%20should%20leave%20the%20current%20version%2C%20after%20our%20discussion.%22%2C%20%22created_at%22%3A%20%222015-09-14T11%3A33%3A38Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/parsing/QueryBuilder.cpp%3AL1-210%22%7D%2C%20%22Pull%20888787134efffdbbabcc4c21dd0cc7479cb578d3%20InformationScripting/src/queries/QueryRegistry.cpp%2063%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/143%23discussion_r39384124%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Move%20this%20to%20some%20AST%20source%20specific%20place.%22%2C%20%22created_at%22%3A%20%222015-09-14T11%3A37%3A17Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/queries/QueryRegistry.cpp%3AL1-113%22%7D%2C%20%22Pull%20778b5e60e4dc3272a27fe47294199c531969399b%20InformationScripting/src/queries/QueryRegistry.cpp%2098%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/143%23discussion_r38790550%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Preferably%2C%20just%20register%20%60%5C%22scripts%5C%22%60%20which%20will%20be%20a%20directory%20in%20DebugBuild.%20Then%20modify%20the%20.pro%20file%20to%20copy%20your%20scripts%20from%20the%20test%20directory%20to%20DebugBuild.%22%2C%20%22created_at%22%3A%20%222015-09-04T20%3A24%3A23Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%2C%20%7B%22body%22%3A%20%22Hm%20I%20can%20do%20that%2C%20but%20then%20it%20gets%20more%20tedious%20to%20edit%20the%20scripts.%20%28you%20have%20to%20move%20manually%20or%20rebuild%29%22%2C%20%22created_at%22%3A%20%222015-09-14T10%3A10%3A11Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/3530769%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/lukedirtwalker%22%7D%7D%2C%20%7B%22body%22%3A%20%22I%20agree%20that%20things%20might%20require%20an%20extra%20step%2C%20but%20it%27s%20the%20correct%20thing%20to%20do.%20Imagine%20a%20release%2C%20binary-only%20distribution%20of%20Envision.%20Then%20there%20should%20be%20a%20directory%20outside%20of%20the%20source%20code%20that%20has%20the%20scripts.%5Cr%5Cn%5Cr%5CnTo%20ease%20debugging%2C%20you%20might%20on%20occasion%20directly%20modify%20the%20script%20in%20DebugBuild%20and%20once%20you%27re%20happy%20with%20it%2C%20just%20copy%20it%20back%20to%20the%20source%20directory.%22%2C%20%22created_at%22%3A%20%222015-09-14T10%3A38%3A25Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/queries/QueryRegistry.cpp%3AL1-113%22%7D%2C%20%22Pull%20778b5e60e4dc3272a27fe47294199c531969399b%20InformationScripting/src/commands/CScript.cpp%2029%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/143%23discussion_r38788697%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Don%27t%20be%20lazy%2C%20add%20the%20%60%3E%200%60%20to%20make%20it%20easier%20to%20read.%22%2C%20%22created_at%22%3A%20%222015-09-04T19%3A58%3A50Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/commands/CScript.cpp%3AL67-204%22%7D%2C%20%22Pull%20778b5e60e4dc3272a27fe47294199c531969399b%20InformationScripting/src/queries/QueryRegistry.h%2047%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/143%23discussion_r38790746%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22You%20have%20declared%2C%20but%20not%20defined%20this%20method.%22%2C%20%22created_at%22%3A%20%222015-09-04T20%3A26%3A57Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/queries/QueryRegistry.h%3AL1-70%22%7D%2C%20%22Pull%20778b5e60e4dc3272a27fe47294199c531969399b%20InformationScripting/src/queries/QueryRegistry.cpp%2065%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov/Envision/pull/143%23discussion_r38790438%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Why%20is%20there%20double%20indentation%20here%3F%22%2C%20%22created_at%22%3A%20%222015-09-04T20%3A22%3A54Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1152976%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/dimitar-asenov%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20InformationScripting/src/queries/QueryRegistry.cpp%3AL1-113%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'><img src='http://www.codereviewhub.com/site/github-bar.png' height=40></a>
- [x] <a href='#crh-comment-Pull 778b5e60e4dc3272a27fe47294199c531969399b InformationScripting/src/commands/CScript.cpp 29'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/143#discussion_r38788697'>File: InformationScripting/src/commands/CScript.cpp:L67-204</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> Don't be lazy, add the `> 0` to make it easier to read.
- [x] <a href='#crh-comment-Pull 778b5e60e4dc3272a27fe47294199c531969399b InformationScripting/src/dataformat/TupleSet.h 17'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/143#discussion_r38789367'>File: InformationScripting/src/dataformat/TupleSet.h:L55-67</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> I'm not quite happy with the name of this method. Let's call it something more descriptive, such as `addPropertiesAsTuples` or some other better name.
- [x] <a href='#crh-comment-Pull 778b5e60e4dc3272a27fe47294199c531969399b InformationScripting/src/parsing/QueryBuilder.cpp 207'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/143#discussion_r38790190'>File: InformationScripting/src/parsing/QueryBuilder.cpp:L1-210</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> OK, I'm not even going to pretend that I really tried to understand the parser. I looked at the code briefly and kept thinking: why is this so complicated? I must be missing something major, because I thought that our command expressions are so simple, we can literally use a few regular expressions to implement the parser. What are you thoughts on this, do we really need all the complex stuff above? If so, could you try to elaborate why that is?
- <a href='https://github.com/lukedirtwalker'><img border=0 src='https://avatars.githubusercontent.com/u/3530769?v=3' height=16 width=16'></a> Hm have you looked at the grammar it should be able to parse in the header file? And after looking at it are you still sure we can solve it with regular expressions? If so I will be happy to redo it with regexes.
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> Ok, I think we should leave the current version, after our discussion.
- [x] <a href='#crh-comment-Pull 778b5e60e4dc3272a27fe47294199c531969399b InformationScripting/src/queries/QueryRegistry.cpp 65'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/143#discussion_r38790438'>File: InformationScripting/src/queries/QueryRegistry.cpp:L1-113</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> Why is there double indentation here?
- [x] <a href='#crh-comment-Pull 778b5e60e4dc3272a27fe47294199c531969399b InformationScripting/src/queries/QueryRegistry.cpp 98'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/143#discussion_r38790550'>File: InformationScripting/src/queries/QueryRegistry.cpp:L1-113</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> Preferably, just register `"scripts"` which will be a directory in DebugBuild. Then modify the .pro file to copy your scripts from the test directory to DebugBuild.
- <a href='https://github.com/lukedirtwalker'><img border=0 src='https://avatars.githubusercontent.com/u/3530769?v=3' height=16 width=16'></a> Hm I can do that, but then it gets more tedious to edit the scripts. (you have to move manually or rebuild)
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> I agree that things might require an extra step, but it's the correct thing to do. Imagine a release, binary-only distribution of Envision. Then there should be a directory outside of the source code that has the scripts.
  To ease debugging, you might on occasion directly modify the script in DebugBuild and once you're happy with it, just copy it back to the source directory.
- [x] <a href='#crh-comment-Pull 778b5e60e4dc3272a27fe47294199c531969399b InformationScripting/src/queries/QueryRegistry.h 39'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/143#discussion_r38790644'>File: InformationScripting/src/queries/QueryRegistry.h:L1-70</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> I have an idea, but I'd like to hear from you, how exactly you imagine we'll use this registry.
- <a href='https://github.com/lukedirtwalker'><img border=0 src='https://avatars.githubusercontent.com/u/3530769?v=3' height=16 width=16'></a> Well it is used to get all known queries in the system. Or what do you mean how we will use it?
- [x] <a href='#crh-comment-Pull 778b5e60e4dc3272a27fe47294199c531969399b InformationScripting/src/queries/QueryRegistry.h 47'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/143#discussion_r38790746'>File: InformationScripting/src/queries/QueryRegistry.h:L1-70</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> You have declared, but not defined this method.
- [x] <a href='#crh-comment-Pull 778b5e60e4dc3272a27fe47294199c531969399b InformationScripting/src/queries/SubstractNodesOperator.cpp 13'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/143#discussion_r38791172'>File: InformationScripting/src/queries/SubstractNodesOperator.cpp:L35-43</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> Hmm, I'm not entirely convinced what's the correct operation of the subtract operator. Strictly speaking, you shouldn't extend the B set before subtracting it from A. But on the other hand, it might be useful, in which case doing it here seems OK, why is it not OK? Maybe it should be an option. I can also imagine that there might be a more complex conditions that determines whether we should subtract something.
- [x] <a href='#crh-comment-Pull 888787134efffdbbabcc4c21dd0cc7479cb578d3 InformationScripting/src/parsing/QueryBuilder.cpp 139'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/143#discussion_r39383850'>File: InformationScripting/src/parsing/QueryBuilder.cpp:L1-210</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> The past tense of split is split.
- [x] <a href='#crh-comment-Pull 888787134efffdbbabcc4c21dd0cc7479cb578d3 InformationScripting/src/queries/QueryRegistry.cpp 63'></a> <img src='http://www.codereviewhub.com/site/github-remaining.png' height=16 width=60>&nbsp;<b><a href='https://github.com/dimitar-asenov/Envision/pull/143#discussion_r39384124'>File: InformationScripting/src/queries/QueryRegistry.cpp:L1-113</a></b>
- <a href='https://github.com/dimitar-asenov'><img border=0 src='https://avatars.githubusercontent.com/u/1152976?v=3' height=16 width=16'></a> Move this to some AST source specific place.

<a href='https://www.codereviewhub.com/dimitar-asenov/Envision/pull/143?mark_as_completed=1'><img src='http://www.codereviewhub.com/site/github-mark-as-completed.png' height=26></a>&nbsp;<a href='https://www.codereviewhub.com/dimitar-asenov/Envision/pull/143?approve=1'><img src='http://www.codereviewhub.com/site/github-approve.png' height=26></a>&nbsp;<a href='https://github.com/dimitar-asenov/Envision/pull/143'><img src='http://www.codereviewhub.com/site/github-refresh.png' height=26></a>
<a href='#crh-end'></a>

lukedirtwalker committed:
Introduce a QueryBuilder to build queries from text.

Per default this parser is now used in the script command.
There is a fallback mode to use hardcoded queries.

lukedirtwalker committed:
Test a color script & add possibility to pass arguments to scripts.

lukedirtwalker committed:
Add a separate QueryRegistry class

lukedirtwalker committed:
Fix small error in NodePropertyAdder

lukedirtwalker committed:
Make test script type safe

lukedirtwalker committed:
Remove condition from NodePropertyAdder

We can achieve the same result by just using filtering before. This
creates a clearer flow of how to do things.

lukedirtwalker committed:
Refactor: Use find in Tuple::operator[]

lukedirtwalker committed:
Update script API and scripts

lukedirtwalker committed:
Allow to add all properties of a cetain type a single tuples.

This can be very helpful for example when substracting nodes where  the
right side has the nodes bound in n-tuples.

dimitar-asenov commented:
This pull request is now managed with CodeReviewHub. <a href='https://github.com/dimitar-asenov/Envision/pull/143'>Refresh</a>.<a href='#crh-update-none'></a>

dimitar-asenov commented on InformationScripting/src/commands/CScript.cpp:
```diff
@@ -65,120 +67,137 @@ Interaction::CommandResult* CScript::execute(Visualization::Item*, Visualization
+
+		if (command == "script")
+		{
+			if (args.size())
```
Don't be lazy, add the `> 0` to make it easier to read.

dimitar-asenov commented on InformationScripting/src/dataformat/TupleSet.h:
```diff
@@ -57,6 +55,12 @@ class INFORMATIONSCRIPTING_API TupleSet
+		 * Adds all properties of type \a T as single Tuples with tag \a tag.
+		 */
+		template<class T>
+		void addAllPropertiesTo(const QString& tag);
```
I'm not quite happy with the name of this method. Let's call it something more descriptive, such as `addPropertiesAsTuples` or some other better name.

dimitar-asenov commented on InformationScripting/src/parsing/QueryBuilder.cpp:
```diff
@@ -0,0 +1,209 @@
+	if (Type::Query == type) return {parseQuery(text)};
+	else if (Type::List == type) return parseList(text);
+	Q_ASSERT(false);
+}
```
OK, I'm not even going to pretend that I really tried to understand the parser. I looked at the code briefly and kept thinking: why is this so complicated? I must be missing something major, because I thought that our command expressions are so simple, we can literally use a few regular expressions to implement the parser. What are you thoughts on this, do we really need all the complex stuff above? If so, could you try to elaborate why that is?

lukedirtwalker replied:
Hm have you looked at the grammar it should be able to parse in the header file? And after looking at it are you still sure we can solve it with regular expressions? If so I will be happy to redo it with regexes.

dimitar-asenov replied:
Ok, I think we should leave the current version, after our discussion.

dimitar-asenov commented on InformationScripting/src/queries/QueryRegistry.cpp:
```diff
@@ -0,0 +1,112 @@
+
+void QueryRegistry::registerDefaultQueries()
+{
+		registerQueryConstructor("classes", [](Model::Node* target, QStringList args) {
```
Why is there double indentation here?

dimitar-asenov commented on InformationScripting/src/queries/QueryRegistry.cpp:
```diff
@@ -0,0 +1,112 @@
+
+void QueryRegistry::registerDefaultScriptLocations()
+{
+	scriptLocations_ << "../InformationScripting/test/scripts/";
```
Preferably, just register `"scripts"` which will be a directory in DebugBuild. Then modify the .pro file to copy your scripts from the test directory to DebugBuild.

lukedirtwalker replied:
Hm I can do that, but then it gets more tedious to edit the scripts. (you have to move manually or rebuild)

dimitar-asenov replied:
I agree that things might require an extra step, but it's the correct thing to do. Imagine a release, binary-only distribution of Envision. Then there should be a directory outside of the source code that has the scripts.

To ease debugging, you might on occasion directly modify the script in DebugBuild and once you're happy with it, just copy it back to the source directory.

dimitar-asenov commented on InformationScripting/src/queries/QueryRegistry.h:
```diff
@@ -0,0 +1,69 @@
+
+class Query;
+
+class INFORMATIONSCRIPTING_API QueryRegistry
```
I have an idea, but I'd like to hear from you, how exactly you imagine we'll use this registry.

lukedirtwalker replied:
Well it is used to get all known queries in the system. Or what do you mean how we will use it?

dimitar-asenov commented on InformationScripting/src/queries/QueryRegistry.h:
```diff
@@ -0,0 +1,69 @@
+		using QueryConstructor = std::function<Query* (Model::Node*, QStringList)>;
+		void registerQueryConstructor(const QString& command, QueryConstructor constructor);
+
+		void registerScriptLocation(const QString& path);
```
You have declared, but not defined this method.

dimitar-asenov commented on InformationScripting/src/queries/SubstractNodesOperator.cpp:
```diff
@@ -33,6 +35,8 @@ QList<TupleSet> SubstractNodesOperator::execute(QList<TupleSet> input)
 	Q_ASSERT(input.size() == 2);
 	auto setA = input[0];
 	auto setB = input[1];
+	// TODO: this is not the correct place to do this. But where is it??
```
Hmm, I'm not entirely convinced what's the correct operation of the subtract operator. Strictly speaking, you shouldn't extend the B set before subtracting it from A. But on the other hand, it might be useful, in which case doing it here seems OK, why is it not OK? Maybe it should be an option. I can also imagine that there might be a more complex conditions that determines whether we should subtract something.

lukedirtwalker committed:
Refactor: Small changes after review

dimitar-asenov commented on InformationScripting/src/parsing/QueryBuilder.cpp:
```diff
@@ -0,0 +1,209 @@
+{
+	// TODO it might be better to be able to register operators.
+	// But that need some additional information on how they are used, thus we currently hardcode the operators.
+	auto splitted = split(text, {'|', '-', 'U'});
```
The past tense of split is split.

dimitar-asenov commented on InformationScripting/src/queries/QueryRegistry.cpp:
```diff
@@ -0,0 +1,112 @@
+	registerDefaultScriptLocations();
+}
+
+void QueryRegistry::registerDefaultQueries()
```
Move this to some AST source specific place.

lukedirtwalker committed:
Only use DebugBuild/scripts directory to search for scripts.

We install the scripts with qmake to this location.

lukedirtwalker committed:
Queries should be registered at initialization of InfoScripting plugin

The QueryRegistry should only contain the information but should not
need to know about the individual queries.

lukedirtwalker committed:
Rename SubstractNodesOperator to SubstractOperator

Since we changed the format to tuples this is now the correct name.
Also don't do anything else than substraction inside.

lukedirtwalker committed:
Rename splitted to splitText

lukedirtwalker committed:
Add a workaround for pre Qt 5.5 compilation
